### DISCUSSION TITLE
Accessing Oqtane DI services/repositories during module install (IInstallable.Install method)

### DISCUSSION THREAD

**Discussion by lanthonyneville** [oqtane/oqtane.framework#4238] [Q&A] [answered]
When my module is installed, I want to automatically create a page in the Admin Dashboard (i.e. with path under admin/ ) that shows it (its a module for showing info about the state of the application to administrators).

So my approach is to implement IModule, IInstallable and in the Install method, use various repository classes to check each site if the page already exists, create it, add module to page etc.

As I am sure this code always runs on the server, I thought it more direct to use the repository classes (e.g. ISiteRepository) than the service classes (e.g. ISiteService). Also there was a useful method in the repository class that was not in the service class - GetPageModules(int siteId).

However, I haven't found a way to get these class instances from DI ... I thought I could do it via static ServiceActivator class like this:

```csharp
var serviceScope = ServiceActivator.GetScope();
var serviceProvider = serviceScope.ServiceProvider;
var siteRepo = serviceProvider.GetRequiredService<ISiteRepository>();
```
... although I get a serviceScope, serviceProvider is null

Is there a way to do it?

One more question - I notice that even when version 1.0.0 of my module is installed, if I install a new version (1.0.1), the Install method is called with version="1.0.0" ... shouldn't it only be called for versions after the last installed version?

**Comment by leigh-pointer** [comment]
You might be able to do this with pagetemplates? When oqtane is created for the first time it uses a pagetemplate class! Search the server project.

**Reply by lanthonyneville** [reply]
Thanks @leigh-pointer - I have now switched to using the PageTemplate class for my auto-generated pages. However, I still need to be able to access 1 DI service to create them: ISiteRepository.CreatePages(Site site, List<PageTemplate> pageTemplates, Alias alias)

I have also realised that my class implementing IInstallable is instantiated by Oqtane using DI, so I can get an IServiceScopeFactory injected in, and from that get an ISiteRepository object.

So this is a lot of progress, but there's one issue still ... when using my ISiteRepository object (I first call GetSites()) the exception below is thrown ... why is this happening (startup has already happened, so services.AddOqtaneDbContext() has already been called) and how can I get around it?

```
System.InvalidOperationException: 'No database provider has been configured for this DbContext. A provider can be configured by overriding the 'DbContext.OnConfiguring' method or by using 'AddDbContext' on the application service provider. If 'AddDbContext' is used, then also ensure that your DbContext type accepts a DbContextOptions<TContext> object in its constructor and passes it to the base constructor for DbContext.'

at Microsoft.EntityFrameworkCore.Internal.DbContextServices.Initialize(IServiceProvider scopedProvider, DbContextOptions contextOptions, DbContext context)
   at Microsoft.EntityFrameworkCore.DbContext.get_ContextServices()
   at Microsoft.EntityFrameworkCore.DbContext.get_Model()
   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.get_EntityType()
   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.CheckState()
   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.get_EntityQueryable()
   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.System.Linq.IQueryable.get_Provider()
   at System.Linq.Queryable.OrderBy[TSource,TKey](IQueryable`1 source, Expression`1 keySelector)
   at Oqtane.Repository.SiteRepository.GetSites() in E:\Projects\Oqtane\oqtane.framework-5.1.1\Oqtane.Server\Repository\SiteRepository.cs:line 109
   at MyCompany.MyProduct.UI.Common.Oqtane.OqtaneModuleDefinition`1.CreateRootPages(Tenant tenant) in E:\TFS\DevSTP1\Architecture\Libraries\MyCompany\MyCompany.Framework.Application.UI\MyCompany.MyProduct.UI.Common\Oqtane\OqtaneModuleDefinition.cs:line 85
   at MyCompany.MyProduct.UI.Common.Oqtane.OqtaneModuleDefinition`1.Install(Tenant tenant, String version) in E:\TFS\DevSTP1\Architecture\Libraries\MyCompany\MyCompany.Framework.Application.UI\MyCompany.MyProduct.UI.Common\Oqtane\OqtaneModuleDefinition.cs:line 73
   at Oqtane.Infrastructure.DatabaseManager.MigrateModules(InstallConfig install) in E:\Projects\Oqtane\oqtane.framework-5.1.1\Oqtane.Server\Infrastructure\DatabaseManager.cs:line 463
   
This exception was originally thrown at this call stack:
	[External Code]
	Oqtane.Repository.SiteRepository.GetSites() in SiteRepository.cs
	... my code calling IInstallable.Install()
	Oqtane.Infrastructure.DatabaseManager.MigrateModules(Oqtane.Shared.InstallConfig) in DatabaseManager.cs
```

**Reply by leigh-pointer** [reply]
Hi @lanthonyneville I would have a look at Oqtane.Infrastructure class NotificationJob in the server project.  There, I know we use IServiceProvider provider to create repositories.

HTH

**Reply by leigh-pointer** [reply]
I this you need this

`var siteRepository = provider.GetRequiredService<ISiteRepository>();`

**Reply by lanthonyneville** [reply]
Yes, that is what I am doing, and I see in NotificationJob the same call. This is my exact code:

```csharp
1. my class implementing IInstallable gets an IServiceScopeFactory injected in
2. var serviceProvider = serviceScopeFactory.CreateScope().ServiceProvider;
3. var siteRepository = serviceProvider.GetRequiredService<ISiteRepository>();
4. var sites = siteRepository.GetSites(); *** Exception here ***
```

**Comment by sbwalker** [comment, marked as answer]
@lanthonyneville I believe "routable modules" is what you should use in this case. Please see this blog post:

https://www.oqtane.org/blog/!/62/routable-modules

You can specify pages to be created within your ModuleInfo.cs class:

```
public class ModuleInfo : IModule
{
    public ModuleDefinition ModuleDefinition => new ModuleDefinition
    {
        Name = "Your Module",
        PageTemplates = new List<PageTemplate>()
        {
            new PageTemplate { AliasName = "*", Path = "admin/yourpage", Parent = "admin", Name = "Your Page" }
        }
    };
} 
```

In regards to the issue with Repository services, you need to remember that Oqtane is multi-tenant and each tenant can have its own database. When a DI Repository service is instantiated dynamically, there is no HttpContext. Without HttpContext the DI Repository service is unable to determine which tenant it should interact with (as tenants are usually determined by the domain name in your HTTP request). So in this case you need to explicitly specify the tenant prior to calling the Repository service.  This is what TenantManager provides:

```
var tenantManager = scope.ServiceProvider.GetRequiredService<ITenantManager>();
tenantManager.SetTenant(TenantId);
// call the Repository service
```

and if you do not know the TenantId you can iterate through the list of tenants (the TenantRepository is tenant-agnostic) so that the logic is executed for every tenant in the installation:

```
var tenantRepository = scope.ServiceProvider.GetRequiredService<ITenantRepository>();
foreach (var tenant in tenantRepository.GetTenants())
{
    tenantManager.SetTenant(tenant.TenantId);
    // call the Repository service
}
```

**Reply by lanthonyneville** [reply]
Thanks @sbwalker for explaining. As you said, using ITenantManager.SetTenant solved the dbcontext issue.

Although routable modules looks easier, I am trying to do it myself via ISiteRepository.CreatePages for various reasons (I am packaging multiple modules in the same assembly as "families", which all share a common area of the nav menu, so want to declare these common page templates in one place in my code, not in every module definition) ....

Going down this road however, I have a new exception, which also looks like a consequence of having no HttpContext ... if you can shed light it would be great ...

ISiteRepository.CreatePages calls _moduleDefinitionRepository.GetModuleDefinitions(site.SiteId), which calls LoadModuleDefinitions(int siteId), which calls ITenantManager.GetAlias() (from which it needs the SiteKey). However this returns null.

Looking into ITenantManager.GetAlias(), if there is no HttpContext it returns _siteState.Alias but only if _siteState.Alias.AliasId != -1 (which it is at this time). I don't know what this means ... _siteState is injected in, but how do I control that?

**Reply by sbwalker** [reply]
@lanthonyneville this path will be challenging due to multi-tenancy. Internally Oqtane stores an Alias object in a scoped service which it uses to determine tenancy. The Alias object contains both a TenantId and SiteId. Most Repository services only require the TenantId so that they can determine the connection string to the tenant database. Other services require the SiteId as well. If you know the SiteId in your code then you can use TenantManager.SetAlias(). That's all I can suggest for now. The routable module concept was designed to simplify the effort for developers.

**Reply by lanthonyneville** [reply]
OK thanks, I'll try that. I can also debug the way routable modules works as there must be the same issue there (no HttpContext). Thanks for all the pointers!

**Reply by lanthonyneville** [reply]
A call to TenantManager.SetAlias did the trick - all my page templates got set up!
